Description: Fixed a bug causing fpdoc to fail silently.
Author: Abou Al Montacir <abou.almontacir@sfr.fr>
Bug-Debian: http://bugs.debian.org/939414
diff --git a/fpcsrc/utils/fpdoc/fpdoc.pp b/fpcsrc/utils/fpdoc/fpdoc.pp
index 01124472..3e07257c 100644
--- a/fpcsrc/utils/fpdoc/fpdoc.pp
+++ b/fpcsrc/utils/fpdoc/fpdoc.pp
@@ -392,6 +392,7 @@ end;
 Procedure TFPDocApplication.DoRun;
 
 begin
+  ExceptionExitCode := 1;
   try
   {$IFDEF Unix}
     gettext.TranslateResourceStrings('/usr/share/locale/%s/LC_MESSAGES/fpdoc-' + {$include %FPCVERSION%} + '.mo');
