diff --git a/fpcsrc/utils/fpcm/fpcmmain.pp b/fpcsrc/utils/fpcm/fpcmmain.pp
index 29e57ff3..934c081d 100644
--- a/fpcsrc/utils/fpcm/fpcmmain.pp
+++ b/fpcsrc/utils/fpcm/fpcmmain.pp
@@ -90,7 +90,7 @@ interface
       );
 
       ppcSuffix : array[TCpu] of string=(
+        '386','68k','ppc','sparc','x64','arm','ppc64','avr','armeb', 'armel', 'mips', 'mipsel', 'mips64', 'mips64el', 'jvm','8086','a64'
-        '386','m68k','ppc','sparc','x64','arm','ppc64','avr','armeb', 'armel', 'mips', 'mipsel', 'mips64', 'mips64el', 'jvm','8086','a64'
       );
 
       OSStr : array[TOS] of string=(
