Description: Prevent FTBFS by creating directories when required
Author: Paul Gevers <elbrus@debian.org>
Index: fpc/fpcdocs/Makefile.fpc
===================================================================
--- fpc.orig/fpcdocs/Makefile.fpc
+++ fpc/fpcdocs/Makefile.fpc
@@ -937,6 +937,7 @@ endif  # USEHEVEA
 fcl.chk: $(FCLXML) fcl-project.xml rtl.chk
         $(FPDOC) $(FPDOCOPTS) --project=fcl-project.xml --format=$(HTMLFMT) --output=fcl$(HTMLSUFFIX) $(FPDOCHTMLOPTS) $(FCLCHMOPTS)
 ifndef CSSFILE
+        mkdir -p fcl
         cp fpdoc.cst fcl/fpdoc.css
 endif
         @$(ECHO) '' > fcl.chk
@@ -944,6 +945,7 @@ endif
 fclres.chk: $(FCLRESXML)
         $(FPDOC) $(FPDOCOPTS) $(FCLRESOPTS)  --format=$(HTMLFMT) --output=fclres$(HTMLSUFFIX) $(FPDOCHTMLOPTS) $(FCLRESCHMOPTS) 
 ifndef CSSFILE
+        mkdir -p fclres
         cp fpdoc.cst fclres/fpdoc.css
 endif
         @$(ECHO) '' > fclres.chk
@@ -951,6 +953,7 @@ endif
 rtl.chk: $(RTLXML) rtl-project.xml
         $(FPDOC) $(FPDOCOPTS) --project=rtl-project.xml --format=$(HTMLFMT) --output=rtl$(HTMLSUFFIX) $(FPDOCHTMLOPTS) $(RTLCHMOPTS)
 ifndef CSSFILE
+        mkdir -p rtl
         cp fpdoc.cst rtl/fpdoc.css
 endif
         @$(ECHO) '' > rtl.chk
